File: 0592-fraction-addition-and-subtraction/0592-fraction-addition-and-subtraction.java
class Solution {
    private int gcd(int a, int b) {
        return a != 0 ? gcd(b % a, a) : Math.abs(b);
    }
    public String fractionAddition(String exp) {
        StringBuilder curNum = new StringBuilder();

        int A = 0, B = 1;
        int a = 0, b = 1;
        
        for (char c : (exp + "+").toCharArray()) {
            if (c == '+' || c == '-') {
                if (curNum.length() > 0) {
                    b = Integer.parseInt(curNum.toString());

                    A = (B * a) + (A * b); B = B * b;

                    int g_c_d = gcd(A, B);
                    A /= g_c_d; B /= g_c_d;
                }
                
                curNum = new StringBuilder().append(c);
            } else if (c == '/') {
                a = Integer.parseInt(curNum.toString());
                curNum = new StringBuilder();
            } else {
                curNum.append(c);
            }
        }

        return A + "/" + B;
    }
}

File: 0592-fraction-addition-and-subtraction/0592-fraction-addition-and-subtraction.py
class Solution:
    def fractionAddition(self, exp: str) -> str:
        fractions = []
        cur_frc, cur_num = [], ""
        
        for c in (exp + "+"):
            if c in ["+","-"]:
                if cur_frc and cur_num:     # for -1/-2
                    cur_frc.append(int(cur_num))
                    fractions.append(cur_frc)
                
                    cur_frc = []
                cur_num = c
            elif c == "/":
                cur_frc.append(int(cur_num))
                cur_num = ""
            else:
                cur_num += c
        
        if not fractions:
            return ""
        
        A,B = fractions[0]
        
        for frac in fractions[1:]:
            n1,n2 = frac
            
            A = (B * n1) + (A * n2)
            B = B * n2
            
            gcd = math.gcd(A, B)
            A, B = (A // gcd), (B // gcd)
            
        
        return "%d/%d" % (A, B)
            